hazelcast/src/main/java/com/hazelcast/instance/impl/Node.java

```diff
@@ -759,25 +774,63 @@ public class NodeShutdownHookThread extends Thread {
         public void run() {
             try {
-                if (isRunning()) {
-                    logger.info("Running shutdown hook... Current state: " + state);
-                    switch (policy) {
-                        case TERMINATE:
-                            hazelcastInstance.getLifecycleService().terminate();
-                            break;
-                        case GRACEFUL:
-                            hazelcastInstance.getLifecycleService().shutdown();
-                            break;
-                        default:
-                            throw new IllegalArgumentException("Unimplemented shutdown hook policy: " + policy);
+                if (!isRunning()) {
+                    return;
+                }
+                final ClusterTopologyIntent shutdownIntent = clusterTopologyIntent.get();
+                if (getNodeExtension().getInternalHotRestartService().isEnabled()
+                        && shutdownIntent != ClusterTopologyIntent.UNKNOWN) {
+                    final ClusterState clusterState = clusterService.getClusterState();
+                    logger.info("Running shutdown hook... Current node state: " + state
+                                + ", detected shutdown intent: " + shutdownIntent
+                                + ", cluster state: " + clusterState);
+                    // consider the detected shutdown intent before triggering node shutdown
+                    if (shutdownIntent == ClusterTopologyIntent.STABLE) {
+                        try {
+                            changeClusterState(ClusterState.FROZEN);
+                        } catch (Throwable t) {
+                            // let shutdown proceed even though we failed to switch to FROZEN state
+                            logger.warning("Could not switch to transient FROZEN state while cluster"
+                                    + "shutdown intent was " + shutdownIntent, t);
+                        }
+                    } else if (shutdownIntent == ClusterTopologyIntent.CLUSTER_SHUTDOWN) {
+                        try {
+                            changeClusterState(ClusterState.PASSIVE);
+                        } catch (Throwable t) {
+                            // let shutdown proceed even though we failed to switch to PASSIVE state
+                            // and wait for replica sync
+                            logger.warning("Could not switch to transient PASSIVE state while cluster"
+                                    + "shutdown intent was " + shutdownIntent, t);
+                        }
+                        if (clusterService.getClusterState() == ClusterState.PASSIVE) {
+                            long timeoutNanos = getProperties().getNanos(ClusterProperty.CLUSTER_SHUTDOWN_TIMEOUT_SECONDS);
+                            getNodeExtension().getInternalHotRestartService()
+                                    .waitPartitionReplicaSyncOnCluster(timeoutNanos, TimeUnit.NANOSECONDS);
```

Comment:
Added a `catch (IllegalStateException e)` that should also cover the case cluster state is not successfully switched to `PASSIVE` (previously it would fail silently)